### PR TITLE
Implement setcnf/applyem

### PR DESCRIPTION
**Pull request by paladine**

### CODE REVIEW

File: MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs [hunk 1]
```diff
@@ -1351,6 +1360,31 @@ public ReadOnlySpan<byte> Invoke(ushort ordinal, bool offsetsOnly = false)
             return null;
         }
 
+        private void setcnf()
+        {
+            string optnam = GetParameterString(0, stripNull: true);
+            string optval = GetParameterString(2, stripNull: true);
+
+            // if updating a new value after an applyem, clear the dictionary and start anew
+            if (applyemCalled)
```
Comment (enusbaum) [inline MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs:1369]: Why wouldn't we do this in `applyem()`? I'm not seeing why we'd defer it to here.

File: MBBSEmu/IO/StreamStream.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+using System;
```
Comment (enusbaum) [inline MBBSEmu/IO/StreamStream.cs:1]: `StreamStream`? 😆

File: MBBSEmu/Module/MsgFile.cs [hunk 3]
```diff
@@ -236,5 +237,81 @@ private void WriteMCV(byte[] language, IList<byte[]> messages, bool writeStringL
             // count of messages
             WriteUInt16(writer, (short)messages.Count);
         }
+
+        public static void UpdateValues(string filename, Dictionary<string, string> values)
+        {
+            var tmpPath = Path.GetTempFileName();
+            StreamStream input = new StreamStream(new FileStream(filename, FileMode.Open));
```
Comment (enusbaum) [inline MBBSEmu/Module/MsgFile.cs:244]: var

File: MBBSEmu/Module/MsgFile.cs [hunk 4]
```diff
@@ -236,5 +237,81 @@ private void WriteMCV(byte[] language, IList<byte[]> messages, bool writeStringL
...
+            StreamStream output = new StreamStream(new FileStream(tmpPath, FileMode.OpenOrCreate));
```
Comment (enusbaum) [inline MBBSEmu/Module/MsgFile.cs:245]: var

File: MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs [hunk 5]
```diff
@@ -1351,6 +1360,31 @@ public ReadOnlySpan<byte> Invoke(ushort ordinal, bool offsetsOnly = false)
             return null;
         }
 
+        private void setcnf()
+        {
+            string optnam = GetParameterString(0, stripNull: true);
```
Comment (enusbaum) [inline MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs:1365]: var

File: MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs [hunk 6]
```diff
@@ -1351,6 +1360,31 @@ public ReadOnlySpan<byte> Invoke(ushort ordinal, bool offsetsOnly = false)
...
+            string optval = GetParameterString(2, stripNull: true);
```
Comment (enusbaum) [inline MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs:1366]: var

File: MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs [hunk 7]
```diff
@@ -1351,6 +1360,31 @@ public ReadOnlySpan<byte> Invoke(ushort ordinal, bool offsetsOnly = false)
...
+                setCnfDictionary.Clear();
+                applyemCalled = false;
+            }
+
+            setCnfDictionary[optnam] = optval;
+        }
+
+        private void applyem()
+        {
+            string filenam = GetParameterString(0, stripNull: true);
```
Comment (enusbaum) [inline MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs:1380]: var

**Review by enusbaum — CHANGES_REQUESTED** [review]

File: MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs (hunk 1 above)
Comment (paladine) [reply MBBSEmu/HostProcess/ExportedModules/Majorbbs.cs:1369]: You can call applyem twice on different files, so if you group all your changes with setcnf you can shard, e.g.

```
setcnf(...);
setcnf(...);
applyem("bbsmajor.msg");
applyem("whatever.msg");

// resets now from scratch
setcnf(...);
```

**Review by enusbaum — APPROVED** [review]

**Pull request merged**